ДЗ 12

morechiba opened the pull request.

morechiba committed:
новая верстка переключателя на темную тему

morechiba committed:
реализация переключения темной темы - сохранение в Shared Preferences

morechiba committed:
поисковая подсказка по фокусу на строке поиска

morechiba committed:
какие-то файлы технические?

morechiba committed:
верстка блока с историей поиска (без точных размеров, нет доступа к Figma)

morechiba committed:
верстка блока с историей поиска с размерами из Figma

morechiba committed:
gradle

morechiba committed:
опять технические какие-то файлы

morechiba committed:
блок истории поиска - прикрутка с MockUp и появлением (не по ТЗ)

morechiba committed:
очистка истории поиска по кнопке
плюс убран плейсхолдер "ничего не найдено" при начале нового поиска

morechiba committed:
onclicklistener на recycler
savetrack при клике
плюс еще что-то

morechiba committed:
savetrack при клике заработал

morechiba committed:
обратный порядок вывода треков  в истории поиска
плюс обновление блока с историей поиска при любых изменениях

morechiba committed:
опять какие-то технические файлы

ArturNurtdinov commented on .idea/AndroidProjectSystem.xml:
Всю папку .idea лучше сразу добавлять в gitignore - в ней хранятся локальные настройки разработчика, специфичные для проекта

ArturNurtdinov commented on app/build.gradle.kts:
```diff
@@ -45,6 +45,8 @@ dependencies {
     implementation("com.google.code.gson:gson:2.10")
     implementation("com.squareup.retrofit2:retrofit:2.11.0")
     implementation("com.squareup.retrofit2:converter-gson:2.11.0")
+    implementation(libs.androidx.recyclerview)
```
Три зависимости выше тоже лучше перевести на использование versions catalog - файл libs.versions.toml

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/App.kt:
```diff
@@ -0,0 +1,50 @@
+
+
+class App: Application() {
+    companion object{
```
`companion object` принято объявлять внизу класса согласно Kotlin Coding Convention (https://kotlinlang.org/docs/coding-conventions.html#class-layout)

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/App.kt:
```diff
@@ -0,0 +1,50 @@
+        const val MODE_NIGHTS = "mode_nights_yes"
+    }
+
+        var darkTheme = false
```
Данной переменной лучше добавить `private set`, чтобы исключить изменение её значения извне, т.к. для изменения значения темы есть функция `switchTheme`

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/SearchActivity.kt:
```diff
@@ -58,29 +60,75 @@ class SearchActivity : AppCompatActivity() {
+        val placeholderNoFound = findViewById<LinearLayout>(R.id.placeholder_no_found)
+        val placeholderNoConnection = findViewById<LinearLayout>(R.id.placeholder_error_connection)
+        val searchHistoryBlock = findViewById<LinearLayout>(R.id.search_history)
+        val SharedPrefs = getSharedPreferences(PLAYLISTMAKER_PREFERENCES, MODE_PRIVATE)
```
Переменные в котлин принято называть с маленьких букв

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,48 @@
+
+    val gson = Gson()
+
+    fun getTracks(): MutableList<Track>{
```
Из функций лучше возвращать просто `List`, а не мьютабельный, чтобы вызывающая сторона не могла модифицировать передаваемые ей данные, а только отрисовывать их

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,48 @@
+    fun getTracks(): MutableList<Track>{
+        val searchHistory: String? = sharedPrefs.getString(SEARCH_HISTORY, "")
+        var trackList: MutableList<Track> = mutableListOf<Track>()
+        if(searchHistory?.isEmpty()!!){
```
Есть функция `isNullOrEmpty()`, чтобы избежать использования оператора `!!`

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,48 @@
+
+        var trackList: MutableList<Track> = gson.fromJson(searchHistory, object : TypeToken<MutableList<Track>>() {}.type) ?: mutableListOf<Track>()
+        trackList.removeIf { it == track }
+        if (trackList.size == 10) {
```
Значение лучше вынести в константу для повышения читабельности кода

ArturNurtdinov commented on app/src/main/java/com/tagomago/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,48 @@
+        var trackList: MutableList<Track> = gson.fromJson(searchHistory, object : TypeToken<MutableList<Track>>() {}.type) ?: mutableListOf<Track>()
+        trackList.removeIf { it == track }
+        if (trackList.size == 10) {
+            trackList.removeAt(0)
```
Удалять необходимо из конца истории поиска, а не начала

morechiba committed:
папка .idea добавлена в .gitignore
плюс companion object перенесен вниз класса согласно Kotlin Coding Convention (https://kotlinlang.org/docs/coding-conventions.html#class-layout)

morechiba committed:
1) переменная с маленькой буквы
2) isNullOrEmpty(), чтобы избежать использования оператора !!
3) Значение для gson.fromJson вынесено в константу для повышения читабельности кода

morechiba committed:
изменение верстки, чтобы кнопка не уходила при 10 треках
плюс исправлено применение isNullOrEmpty()

morechiba committed:
другой порядок сохранения треков в истории
